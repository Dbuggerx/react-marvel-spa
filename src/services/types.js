// @flow

export type Hero = {
    id: number,
    name: string,
    description: string,
    modified: Date,
    resourceURI: string,
    urls: [
        {
            type: string,
            url: string,
        },
    ],
    thumbnail: {
        path: string,
        extension: string,
    },
    comics: {
        available: number,
        returned: number,
        collectionURI: string,
        items: [
            {
                resourceURI: string,
                name: string,
            },
        ],
    },
    stories: {
        available: number,
        returned: number,
        collectionURI: string,
        items: [
            {
                resourceURI: string,
                name: string,
                type: string,
            },
        ],
    },
    events: {
        available: number,
        returned: number,
        collectionURI: string,
        items: [
            {
                resourceURI: string,
                name: string,
            },
        ],
    },
    series: {
        available: number,
        returned: number,
        collectionURI: string,
        items: [
            {
                resourceURI: string,
                name: string,
            },
        ],
    },
}

export type HeroesResult = {
    code: number,
    status: string,
    copyright: string,
    attributionText: string,
    attributionHTML: string,
    data: {
        offset: number,
        limit: number,
        total: number,
        count: number,
        results: Hero[],
    },
    etag: string,
};
