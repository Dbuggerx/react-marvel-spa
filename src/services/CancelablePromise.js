// @flow

export default class CancelablePromise<T> {
    constructor(promise: Promise<T>) {
        this.promise = new Promise((resolve, reject) => {
            promise.then(
                (val: T) => (this.canceled ? reject(new Error(this.constructor.errorMessage)) : resolve(val)),
                (error: Error) => (this.canceled ? reject(new Error(this.constructor.errorMessage)) : reject(error)),
            );
        });
    }
    promise: Promise<T>;
    canceled = false;

    cancel() {
        this.canceled = true;
    }

    static errorMessage = 'canceled';
}
