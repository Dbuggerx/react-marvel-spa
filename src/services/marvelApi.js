// @flow
import type { HeroesResult, Hero } from './types';

const apiKey = '570b1cdca67e5243b4db4425b1b4ee4e';
const imageSizes = {
    small: 'standard_medium',
    big: 'portrait_uncanny',
};

async function getJsonFromUrl(url: string) {
    const response = await fetch(url);
    if (!response.ok) {
        throw new Error(response.statusText);
    }
    return response.json();
}

export async function fetchHeros(offset: number = 0): Promise<HeroesResult> {
    const url = `https://gateway.marvel.com:443/v1/public/characters?offset=${offset}&apikey=${apiKey}`;
    return getJsonFromUrl(url);
}

export async function fetchHeroById(id: number): Promise<HeroesResult> {
    const url = `https://gateway.marvel.com:443/v1/public/characters/${id}/?apikey=${apiKey}`;
    return getJsonFromUrl(url);
}

export async function fetchHeroSearch(nameStartsWith: string, offset: number = 0): Promise<HeroesResult> {
    const url = `https://gateway.marvel.com:443/v1/public/characters?nameStartsWith=${nameStartsWith}&offset=${offset}&apikey=${apiKey}`;
    return getJsonFromUrl(url);
}

export function getImageUrl(hero: Hero, size: $Keys<typeof imageSizes>) {
    return `${hero.thumbnail.path}/${imageSizes[size]}.${hero.thumbnail.extension}?apikey=${apiKey}`;
}
