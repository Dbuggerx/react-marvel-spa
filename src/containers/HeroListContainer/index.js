// @flow

import React, { Component } from 'react';
import { fetchHeros, getImageUrl } from '../../services/marvelApi';
import ErrorMessage from '../../components/ErrorMessage';
import LoadingIndicator from '../../components/LoadingIndicator';
import HeroList from '../../components/HeroList';
import CancelablePromise from '../../services/CancelablePromise';
import type { Hero } from '../../services/types';

type State = {
    isLoading: boolean,
    heroes: Hero[],
    error?: string,
};

export default class HeroListContainer extends Component<void, State> {
    constructor(props: void) {
        super(props);
        this.state = {
            heroes: [],
            isLoading: true,
        };
    }

    componentDidMount() {
        this.fetchMarvelHeroes();
    }

    componentWillUnmount() {
        if (this.cancelableFetch) {
            this.cancelableFetch.cancel();
        }
    }

    getThumbnailUrl = (hero: Hero) => getImageUrl(hero, 'small');

    cancelableFetch: *;

    async fetchMarvelHeroes() {
        try {
            this.cancelableFetch = new CancelablePromise(fetchHeros());
            const heroesResult = await this.cancelableFetch.promise;
            this.setState({
                heroes: heroesResult.data.results,
                isLoading: false,
            });
        } catch (error) {
            if (error.message !== CancelablePromise.errorMessage) {
                this.setState({
                    error: 'Error loading heroes',
                });
            }
        }
    }

    render() {
        if (this.state.error) {
            return <ErrorMessage message={this.state.error} />;
        }
        if (this.state.isLoading) {
            return <LoadingIndicator />;
        }
        return (
            <div>
                <HeroList heroes={this.state.heroes} getThumbnailUrl={this.getThumbnailUrl} />
            </div>
        );
    }
}
