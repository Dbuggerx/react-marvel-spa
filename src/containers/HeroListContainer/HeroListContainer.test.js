import React from 'react';
import TestRenderer from 'react-test-renderer';
import HeroListContainer from './index';
import { fetchHeros, getImageUrl } from '../../services/marvelApi';

jest.mock('../../services/marvelApi');

describe('HeroListContainer', () => {
    beforeEach(() => {
        fetchHeros.mockClear();
        getImageUrl.mockClear();
    });

    test('it renders a loading message', () => {
        // A simple mocked Promise will do to test the loading message
        fetchHeros.mockReturnValue(Promise.resolve());

        const renderer = TestRenderer.create(<HeroListContainer />);
        expect(renderer.toJSON()).toMatchSnapshot();
        renderer.unmount();
    });

    test('it renders an error message', (done) => {
        // Simulate an error
        fetchHeros.mockReturnValue(Promise.reject(new Error('Testing an error')));

        const container = <HeroListContainer />;
        const renderer = TestRenderer.create(container);

        process.nextTick(() => {
            renderer.update(container);
            expect(renderer.toJSON()).toMatchSnapshot();
            renderer.unmount();
            done();
        });
    });

    test('it renders heroes', (done) => {
        // Mock the response from the API
        fetchHeros.mockReturnValue(Promise.resolve({
            code: 200,
            status: 'Ok',
            data: {
                offset: 0,
                limit: 2,
                total: 1491,
                count: 2,
                results: [
                    {
                        id: 1011334,
                        name: '3-D Man',
                        description: '',
                        modified: '2014-04-29T14:18:17-0400',
                        thumbnail: {
                            path: 'http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784',
                            extension: 'jpg',
                        },
                    },
                    {
                        id: 1017100,
                        name: 'A-Bomb (HAS)',
                        description:
                                "Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction! ",
                        modified: '2013-09-18T15:54:04-0400',
                        thumbnail: {
                            path: 'http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16',
                            extension: 'jpg',
                        },
                    },
                ],
            },
        }));

        // Using the real implementation for 'getImageUrl'
        getImageUrl.mockImplementation(require.requireActual('../../services/marvelApi').getImageUrl);

        const container = <HeroListContainer />;
        const renderer = TestRenderer.create(container);

        process.nextTick(() => {
            renderer.update(container);
            expect(renderer.toJSON()).toMatchSnapshot();
            renderer.unmount();
            done();
        });
    });
});
