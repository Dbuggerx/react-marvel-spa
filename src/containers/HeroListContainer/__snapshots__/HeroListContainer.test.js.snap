// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroListContainer it renders a loading message 1`] = `
<h1>
  Loading...
</h1>
`;

exports[`HeroListContainer it renders an error message 1`] = `
<h1>
  Error loading heroes
</h1>
`;

exports[`HeroListContainer it renders heroes 1`] = `
<div>
  <div>
    <div>
      <img
        alt="3-D Man"
        src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784/standard_medium.jpg?apikey=570b1cdca67e5243b4db4425b1b4ee4e"
      />
      Name: 
      3-D Man
    </div>
    <div>
      <img
        alt="A-Bomb (HAS)"
        src="http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16/standard_medium.jpg?apikey=570b1cdca67e5243b4db4425b1b4ee4e"
      />
      Name: 
      A-Bomb (HAS)
    </div>
  </div>
</div>
`;
